java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -0,0 +1,376 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.se;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.java.cfg.CFG;
+import org.sonar.java.model.JavaTree;
+import org.sonar.java.se.checkers.ConditionAlwaysTrueOrFalseChecker;
+import org.sonar.java.se.checkers.NullDereferenceChecker;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import javax.annotation.CheckForNull;
+
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+public class ExplodedGraphWalker extends BaseTreeVisitor {
+
+  /**
+   * Arbitrary number to limit symbolic execution.
+   */
+  private static final int MAX_STEPS = 2000;
+  private static final Logger LOG = LoggerFactory.getLogger(ExplodedGraphWalker.class);
+  private static final Set<String> THIS_SUPER = ImmutableSet.of("this", "super");
+
+  private final ConditionAlwaysTrueOrFalseChecker alwaysTrueOrFalseChecker;
+  private ExplodedGraph explodedGraph;
+  private Deque<ExplodedGraph.Node> workList;
+  private ExplodedGraph.Node node;
+  public ExplodedGraph.ProgramPoint programPosition;
+  public ProgramState programState;
+
+  private CheckerDispatcher checkerDispatcher;
+
+  @VisibleForTesting
+  int steps;
+  ConstraintManager constraintManager;
+
+  public static class MaximumStepsReachedException extends RuntimeException {
+    public MaximumStepsReachedException(String s) {
+      super(s);
+    }
+  }
+
+  public ExplodedGraphWalker(JavaFileScannerContext context) throws MaximumStepsReachedException {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove the declaration of thrown exception 'org.sonar.java.se.ExplodedGraphWalker$MaximumStepsReachedException' which is a runtime exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)